stl/inc/format

```diff
@@ -0,0 +1,2734 @@
+    // At most one codepoint (so one char32_t or four utf-8 char8_t).
+    _CharT _Fill[4 / sizeof(_CharT)] = {_CharT{' '}};
+};
+
+// Adds width and precision references to _Basic_format_specs.
+// this is required for std::formatter implementations because we must
+// parse the format specs without having access to the format args (via a format context)
+template <class _CharT>
+struct _Dynamic_format_specs : _Basic_format_specs<_CharT> {
+    int _Dynamic_width_index     = -1;
+    int _Dynamic_precision_index = -1;
+};
+
+// Model of _Parse_spec_callbacks that fills a _Basic_format_specs with the parsed data.
+template <class _CharT>
+class _Specs_setter {
+public:
+    explicit constexpr _Specs_setter(_Basic_format_specs<_CharT>& _Specs_) : _Specs(_Specs_) {}
+
+    constexpr void _On_align(_Align _Aln) {
+        _Specs._Alignment = _Aln;
+    }
+
+    constexpr void _On_fill(basic_string_view<_CharT> _Sv) {
+        if (_Sv.size() > _STD size(_Specs._Fill)) {
+            throw format_error("Invalid fill (too long).");
+        }
+
+        const auto _Pos = _STD _Copy_unchecked(_Sv._Unchecked_begin(), _Sv._Unchecked_end(), _Specs._Fill);
+        _STD fill(_Pos, _STD end(_Specs._Fill), _CharT{});
+        _Specs._Fill_length = static_cast<uint8_t>(_Sv.size());
+    }
+
+    constexpr void _On_sign(_Sign _Sgn) {
+        _Specs._Sgn = _Sgn;
+    }
+
+    constexpr void _On_hash() {
+        _Specs._Alt = true;
+    }
+
+    constexpr void _On_zero() {
+        _Specs._Leading_zero = true;
+    }
+
+    constexpr void _On_width(int _Width) {
+        _Specs._Width = _Width;
+    }
+
+    constexpr void _On_precision(int _Precision) {
+        _Specs._Precision = _Precision;
+    }
+
+    constexpr void _On_localized() {
+        _Specs._Localized = true;
+    }
+
+    constexpr void _On_type(_CharT _Type) {
+        // performance note: this could be optimized to one comparison by
+        // first casting to unsigned int (the negative values will be 128-255)
+        if (_Type < 0 || _Type > (numeric_limits<signed char>::max)()) {
+            throw format_error("Invalid type specification.");
+        }
+        _Specs._Type = static_cast<char>(_Type);
+    }
+
+protected:
+    _Basic_format_specs<_CharT>& _Specs;
+};
+
+template <class _Context>
+_NODISCARD constexpr basic_format_arg<_Context> _Get_arg(const _Context& _Ctx, size_t _Arg_id) {
+    // note: while this is parameterized on the _Arg_id type in libfmt we don't
+    // need to do that in std::format because it's only called with either an integer
+    // id or a named id (which we do not support in std::format)
+    auto _Arg = _Ctx.arg(_Arg_id);
+    if (!_Arg) {
+        throw format_error("Argument not found.");
+    }
+    return _Arg;
+}
+
+// Checks that the type and value of an argument associated with a dynamic
+// width specifier are valid.
+class _Width_checker {
+public:
+    template <class _Ty>
+    _NODISCARD constexpr unsigned long long operator()(_Ty _Value) {

```

Comment:
Should we use `auto` here given that we cast below. Should that be a const method?
```suggestion
    _NODISCARD constexpr auto operator()(const _Ty _Value) const {
```

Reply:
We prefer to use `auto` when initializing a variable from a cast, since there's no reason to repeat the type information on a single line. However, I don't believe that this should be carried over to a function's return type, where `auto` forces readers to look at the implementation. (`auto` and `decltype(auto)` are useful when the return type is being computed in a nontrivial manner.)

Adding `const` is good though.